src/DistIL/IR/DSL/ValueMatchInterpolator.cs

```diff
@@ -0,0 +1,59 @@
+namespace DistIL.IR.DSL;
+
+using System.Runtime.CompilerServices;
+
+[InterpolatedStringHandler]
+public unsafe struct ValueMatchInterpolator(int literalLength, int formattedCount)
+{
+    public readonly Dictionary<string, IntPtr> Outputs = [];
+    public readonly Dictionary<string, Value> OutputBuffer = [];
+    private readonly StringBuilder _builder = new StringBuilder();
+
+    public void AppendLiteral(string value)
+    {
+        _builder.Append(value);
+    }
+
+    public void AppendFormatted<T>(in T value, [CallerArgumentExpression("value")] string? name = null)
+        where T : Value
+    {
+        if (name != null)
+        {
+            Outputs[name] = (IntPtr)Unsafe.AsPointer(ref Unsafe.AsRef(in value));

```

Comment:
Oh no. This should never be done! `Unsafe.AsPointer` is the most dangerous method on `Unsafe` class because this is guaranteed to introduce random crashes because it is illegal to reinterpret a byref that points to managed memory without pinning that said memory. It's unfortunate it does not come with a compiler warning/error.

(otherwise thank you for contributing to DistIL but oh no, not the Unsafe.AsPointer!

Reply:
Do you have an idea how to implement the output of values else?

Reply:
My reasoning here was that this method is only supposed to be used with local variables, which are guaranteed to not be moved around by the GC, and where I expect this should be mostly fine.

It doesn't change the fact that it's a scary hack, but the long-term plan is to use a source interceptor to replace the interpolation handlers with specialized matchers to avoid the overhead from parsing/matching.

Reply:
Got it. The issue with byrefs is they can e.g. point to variables lifted into a display class by a closure, which means they would no longer point to stack. Byrefs are also necessary to keep objects alive - pointers will not and because .NET has precise GC tracking it means that the array if pointers to *stack locations* is insufficient to keep objects stored at those locations alive.

Reply:
Thanks for the info, I was not aware of these problems. We'll probably have to look for another solution to get the output values then.

> pointers to stack locations is insufficient to keep objects stored at those locations alive

Just to clarify, does this mean that `object x; *(&x) = new();` (e.g. a roundtrip through ldloca/conv.u/stind) is effectively a GC hole and it's only safe to store managed objects to refs/tracked pointers?

Reply:
> Just to clarify, does this mean that object x; *(&x) = new(); (e.g. a roundtrip through ldloca/conv.u/stind) is effectively a GC hole and it's only safe to store managed objects to refs/tracked pointers?

Yes, even though the simplified snippet like this won't trigger it. The above example is more complex and is definitely a GC hole. Even the snippet itself - if this local is the last place where it is observed, then the GC will be free to collect the objects since there is no object or managed reference to keep it alive.

By reinterpreting byrefs into pointers it completely breaks the tracking and allows GC to collect `Value` objects the `T` here is constrained on. This could theoretically be less problematic for `T` satisfying `unmanaged` constraint but it does not address a scenario where a byref points to a field on the heap.

Reply:
What about something like this?

```csharp
if (instr.Match("(add {x} (mul ? {y}))", out var outputs) {
    var x = outputs.Get<ConstInt>("x");
    var y = outputs.Get<Instruction>("y");
}
```